pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -144,23 +147,70 @@ type stringSliceFunc func(storage.Querier) ([]string, storage.Warnings, error)
 // the output of the stringSliceFunc to be sorted, as results of LabelValues
 // are not sorted.
 //
-// TODO: Consider running stringSliceFunc calls concurrently
+
+type stringSliceFuncResult struct {
+	Result   []string
+	Warnings storage.Warnings
+}
+
 func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, storage.Warnings, error) {
+
+	g, _ := errgroup.WithContext(m.ctx)
+
 	var warnings storage.Warnings
 	resultMap := make(map[string]struct{})
-	for pos, tenantID := range m.tenantIDs {
-		result, resultWarnings, err := f(m.queriers[pos])
-		if err != nil {
-			return nil, nil, err
-		}
-		for _, e := range result {
-			resultMap[e] = struct{}{}
-		}
-		for _, w := range resultWarnings {
-			warnings = append(warnings, fmt.Errorf("error querying tenant id %s: %w", tenantID, w))
+	resultCh := make(chan *stringSliceFuncResult)
+	resultDone := make(chan struct{})
+
+	go func() {
+		defer func() {
+			close(resultDone)
+		}()
+		for result := range resultCh {
+			for _, e := range result.Result {
+				resultMap[e] = struct{}{}
+			}
+			warnings = append(warnings, result.Warnings...)
 		}
+	}()
+
+	for pos := range m.tenantIDs {
+		querier := m.queriers[pos]
+		tenantID := m.tenantIDs[pos]
+		g.Go(func() error {
+			result, resultWarnings, err := f(querier)
+			if err != nil {
+				return errors.Wrapf(err, `error querying {%s="%s"}`, defaultTenantLabel, tenantID)
+			}
+
+			var warnings = make(storage.Warnings, len(resultWarnings))
+			for pos := range resultWarnings {
+				warnings[pos] = errors.Wrapf(resultWarnings[pos], `warning querying {%s="%s"}`, defaultTenantLabel, tenantID)
+			}
+
+			resultCh <- &stringSliceFuncResult{
+				Result:   result,
+				Warnings: warnings,
+			}
+
+			return nil
+		})
+	}
+
+	// await first error or return of all results
```

Comment:
This comment is slightly misleading -- it indicates that it doesn't wait for all goroutines to stop first. Fortunately it does, otherwise `close` on next line could cause goroutines to panic, when trying to write to closed channel.

Reply:
Fixed in a5b94abc6e002f6038d755fc1e9007316b3c3712